### PR TITLE
SL - Amanda Scott  

### PR DESCRIPTION
**Pull request by ascott43517**

### CODE REVIEW

File: test/adagrams.test.js [hunk 1]
```diff
@@ -120,7 +120,9 @@ describe("Adagrams", () => {
     });
 
     it("returns a score of 0 if given an empty input", () => {
-      throw "Complete test";
+      expectScores({
```
Comment (anselrognlie) [inline test/adagrams.test.js:123]: 👍

File: test/adagrams.test.js [hunk 2]
```diff
@@ -133,19 +135,21 @@ describe("Adagrams", () => {
...
       expect(highestScoreFrom(words)).toEqual(correct);
+      
     });
 
     it("accurately finds best scoring word even if not sorted", () => {
       const words = ["XXX", "XXXX", "X", "XX"];
       const correct = { word: "XXXX", score: scoreWord("XXXX") };
 
-      throw "Complete test by adding an assertion";
+      expect(highestScoreFrom(words)).toEqual(correct);
```
Comment (anselrognlie) [inline test/adagrams.test.js:152]: 👍

File: src/adagrams.js [hunk 3]
```diff
@@ -1,15 +1,134 @@
 export const drawLetters = () => {
-  // Implement this method for wave 1
+  const letterPool = {
```
Comment (anselrognlie) [inline src/adagrams.js:2]: Consider moving this outside the function (or wrap it aggressively) so that the function becomes shorter. Depending on our approach, we may need to copy the data each function call. We don't need to with the way this is currently written since the pool is recreated fresh each time the function is called.

File: src/adagrams.js [hunk 4]
```diff
@@ -1,15 +1,134 @@
...
+    W : 2,
+    K : 1,
+    X : 1,
+    L : 4,    
+    Y : 2, 
+    M : 2,
+    Z : 1,   
+  }
+
+let newPool = letterPool
```
Comment (anselrognlie) [inline src/adagrams.js:31]: We don't need a new variable here. `letterPool` is created fresh each time the function is called, so we can freely modify `letterPool` without issue.

Also, this wouldn't make a copy. `newPool` refers to the same data structure as `letterPool`. To copy, we would need to make an explicit copy using something like `assign` or using spread syntax.

File: src/adagrams.js [hunk 5]
```diff
@@ -1,15 +1,134 @@
...
+for (let i = 0; i < 10; i++){
+  let randomIndex = Math.floor(Math.random() * 26)
+  let key = Object.keys(newPool)[randomIndex]
+  if (!hand.includes(key)){
+    newPool[key] -= 1
+    hand.push(key);
+  } else if (hand.includes(key) && newPool[key] > 0){
+    newPool[key] -= 1
+    hand.push(key)
+  }
```
Comment (anselrognlie) [inline src/adagrams.js:42]: Note that this doesn't quite reflect the same proability distribution of if all the tiles were in one big pile and we randomly picked 10.  This will produce a higher likelihood of getting "rare" letters.

Your logic does properly ensure that you don't give more letters than are available.

File: src/adagrams.js [hunk 6]
```diff
@@ -1,15 +1,134 @@
...
+    Y : 2, 
+    M : 2,
+    Z : 1,   
+  }
+
+let newPool = letterPool
+let hand = []
+for (let i = 0; i < 10; i++){
+  let randomIndex = Math.floor(Math.random() * 26)
+  let key = Object.keys(newPool)[randomIndex]
```
Comment (anselrognlie) [inline src/adagrams.js:35]: Clever way to pick one of the letters. This wouldn't work in python, since we can't use `[]` on the `keys` result. But JS returns an actual array, so for JS it works.

Comment (anselrognlie) [reply src/adagrams.js:35]: Consider making the `Object.keys` call only once, outside the loop. It really does have to tierate over the object and pack the keys into a dictionary, which depends on the size of the object. Since the available keys don't change over the course of the function, we can call it once and store it in a variable to use each time through the loop.

File: src/adagrams.js [hunk 7]
```diff
@@ -1,15 +1,134 @@
...
+  }
 };
 
 export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+  let letterArray = []
+  for (const letter of input){
+    
+    letterArray.push(letter);
+  }
```
Comment (anselrognlie) [inline src/adagrams.js:54]: I was trying to figure out why you were packing input into an array until I remembered that `input` is a string, not an array. We should have picked a better parameter name! 😆 

That being said, rather than writing this out ourselves, we could use one of the following methods to turn the string into an array
```js
input.split('');
[...input];
Array.from(input);
```

File: src/adagrams.js [hunk 8]
```diff
@@ -1,15 +1,134 @@
...
+    for (const letter of input) {
+    const letterCountInHand = lettersInHand.filter(x=>x===letter).length
+    const letterCountInInput = letterArray.filter(x=>x===letter).length
```
Comment (anselrognlie) [inline src/adagrams.js:57]: Filtering the lists to get their length (counting the number of copies) requires iterating through the lists for each input. So we're doing a nested loop. Consider building frequency maps for the input and the hand which requires iterating over each only once. Then we could compare the counts as you're doing below.

File: src/adagrams.js [hunk 9]
```diff
@@ -1,15 +1,134 @@
...
+    if (letterCountInInput === letterCountInHand){
+      continue;
+    } else {
+      return false;
+    }
```
Comment (anselrognlie) [inline src/adagrams.js:62]: This check is overly strict. They don't need the _same_ quantity. We only need the hand count to be greater or equal to the input count. Prefer to avoid condition blocks solely containing a continue when there's no other code being skipped. Taken together, we could write this as:
```js
    if (letterCountInInput > letterCountInHand){
      return false;
    }   
```

File: src/adagrams.js [hunk 10]
```diff
@@ -1,15 +1,134 @@
...
+    if (letterCountInInput === letterCountInHand){
+      continue;
+    } else {
+      return false;
+    }   
+}
+return true;
 };
 
 export const scoreWord = (word) => {
```
Comment (anselrognlie) [inline src/adagrams.js:67]: 👍

File: src/adagrams.js [hunk 11]
```diff
@@ -1,15 +1,134 @@
...
-  // Implement this method for wave 3
+  const scoreChart = {
```
Comment (anselrognlie) [inline src/adagrams.js:68]: Consider moving this outside the function (or wrap it aggressively) so that the function becomes shorter.

File: src/adagrams.js [hunk 12]
```diff
@@ -1,15 +1,134 @@
...
+    'word': 0,
+    'score': 0
+  }
+  let scoredWords = [];
+  let highest = 0;
+  for (const word of words){
+    let scores = scoreWord(word);
+    if (scores > highest){
+      highest = scores;
+      scoredWords.push(scores);
```
Comment (anselrognlie) [inline src/adagrams.js:119]: `scoredWords` doesn't appear to be used for anything.

File: src/adagrams.js [hunk 13]
```diff
@@ -1,15 +1,134 @@
...
+      winningWord.word = word;
+      winningWord.score = scores;
+    } 
+    if (scores === highest && word.length === 10){
+      winningWord.word = word;
+    }
+    if (scores === highest && word.length && winningWord.word.length === 10){
+      break;
+    }
```
Comment (anselrognlie) [inline src/adagrams.js:128]: I'm not sure about this condition. It says if the current word's score ties the highest, and the current word is not 0-length (`word.length`) and the current winning word is 10-letters, break out (which results in `winningWord` being returned). But I don't see why it would be safe to `break` out here (`highest` is only the highest we've seen so far, not necessarily the absolue highest) and I also don't understand why the non-zero length check is here. What was the intent of this check?

File: src/adagrams.js [hunk 14]
```diff
@@ -1,15 +1,134 @@
...
+      winningWord.word = word;
+      winningWord.score = scores;
+    } 
+    if (scores === highest && word.length === 10){
+      winningWord.word = word;
+    }
+    if (scores === highest && word.length && winningWord.word.length === 10){
+      break;
+    } 
+    if (scores === highest && word.length < winningWord.word.length) {
```
Comment (anselrognlie) [inline src/adagrams.js:129]: I feel like most of these `if`s should `else if`s so that only one of them takes effect each loop. For instance, for this final check, we wouldn't want a shorter word to replace the winning word if winning word's length were 10, but I think this code would allow that to happen with specific inputs.

**Review by anselrognlie — COMMENTED** [review]
Looks good! I see a healthy mixture of code that might feel more at home in Python, but also some approaches that feel very JS. When using more functional JS features, remember to keep in mind where the function call might be hiding a loop, and try to avoid using those within an outer loop. Even though the small size of the data in this means the absolute performance is still fast, we should keep the algorithmic complexity in mind. Imagine what would happen if we relaxed the data restrictions to use larger tile sets or hands. Your tests are all passing, but take a look at my comments on Wave 4. There might be a corner case to revisit. Overall, nicely done!
